nixos/tests/installer.nix

```diff
@@ -68,160 +67,180 @@ let
+        qemuFlags =
+          (if system == "x86_64-linux" then "-m 768 " else "-m 512 ") +
+          (optionalString (system == "x86_64-linux") "-cpu kvm64 ") +
+          (optionalString (system == "aarch64-linux") "-enable-kvm -machine virt,gic-version=host -cpu host ");
+
+        hda = "vm-state-machine/machine.qcow2";
+        hdaInterface = iface;
+      } // (optionalAttrs isEfi {
+        bios = if pkgs.stdenv.isAarch64
+          then "${pkgs.OVMF.fd}/FV/QEMU_EFI.fd"
+          else "${pkgs.OVMF.fd}/FV/OVMF.fd";
+      })));
 
-      hdFlags = ''hda => "vm-state-machine/machine.qcow2", hdaInterface => "${iface}", ''
-        + optionalString isEfi (if pkgs.stdenv.isAarch64
-            then ''bios => "${pkgs.OVMF.fd}/FV/QEMU_EFI.fd", ''
-            else ''bios => "${pkgs.OVMF.fd}/FV/OVMF.fd", '');
     in if !isEfi && !(pkgs.stdenv.isi686 || pkgs.stdenv.isx86_64) then
       throw "Non-EFI boot methods are only supported on i686 / x86_64"
     else ''
+      def create_machine_from_file(machine_config_file):
+          with open(machine_config_file, "r") as f:
+              return create_machine(json.load(f))
+
 
-      $machine->start;
+      machine.start()
 
       # Make sure that we get a login prompt etc.
-      $machine->succeed("echo hello");
-      #$machine->waitForUnit('getty@tty2');
-      #$machine->waitForUnit("rogue");
-      $machine->waitForUnit("nixos-manual");
+      machine.succeed("echo hello")
+      # machine.wait_for_unit("getty@tty2")
+      # machine.wait_for_unit("rogue")
+      machine.wait_for_unit("nixos-manual")
 
       # Wait for hard disks to appear in /dev
-      $machine->succeed("udevadm settle");
+      machine.succeed("udevadm settle")
 
       # Partition the disk.
       ${createPartitions}
 
       # Create the NixOS configuration.
-      $machine->succeed("nixos-generate-config --root /mnt");
+      machine.succeed("nixos-generate-config --root /mnt")
 
-      $machine->succeed("cat /mnt/etc/nixos/hardware-configuration.nix >&2");
+      machine.succeed("cat /mnt/etc/nixos/hardware-configuration.nix >&2")
 
-      $machine->copyFileFromHost(
+      machine.copy_file_from_host(
           "${ makeConfig { inherit bootLoader grubVersion grubDevice grubIdentifier grubUseEfi extraConfig; } }",
-          "/mnt/etc/nixos/configuration.nix");
+          "/mnt/etc/nixos/configuration.nix",
+      )
 
       # Perform the installation.
-      $machine->succeed("nixos-install < /dev/null >&2");
+      machine.succeed("nixos-install < /dev/null >&2")
 
       # Do it again to make sure it's idempotent.
-      $machine->succeed("nixos-install < /dev/null >&2");
+      machine.succeed("nixos-install < /dev/null >&2")
 
-      $machine->succeed("umount /mnt/boot || true");
-      $machine->succeed("umount /mnt");
-      $machine->succeed("sync");
+      machine.succeed("umount /mnt/boot || true")
+      machine.succeed("umount /mnt")
+      machine.succeed("sync")
```

Comment:
all these groups of steps would be perfect candidates for descriptive `with subtest(....):` sections.

Reply:
yeah i thought so to. just wanted to do a straight port first without refactoring at the same time.

Reply:
We can do a followup PR, this shouldn't be a reason to block this PR.